### PR TITLE
Refactor KvRange

### PR DESCRIPTION
**Pull request by nrc**
The end result is a concrete type `BoundRange` and using `Into<BoundRange>`
where functions previously took `KeyRange`. This refactoring means that errors
with invalid ranges are found earlier (in fact, they are mostly prevented statically,
where they can still occur, they do so when ranges are created, not when used)
and there is less conversion between range types in some cases. Hopefully,
code is simplified somewhat too.

PTAL @Hoverbear @brson @sunxiaoguang

### CODE REVIEW

File: src/kv.rs [hunk 1]
```diff
@@ -320,171 +350,179 @@ impl fmt::Debug for KvPair {
...
+        // Debug assert because this function is private.
+        debug_assert!(from != Bound::Unbounded);
+        BoundRange { from, to }
+    }
+
     /// Ranges used in scanning TiKV have a particularity to them.
     ///
     /// The **start** of a scan is inclusive, unless appended with an '\0', then it is exclusive.
     ///
     /// The **end** of a scan is exclusive, unless appended with an '\0', then it is inclusive.
```
Comment (Hoverbear) [inline src/kv.rs]: Perhaps we should be noting this function adds these `\0`s?

File: src/rpc/client.rs [hunk 2]
```diff
@@ -373,14 +374,14 @@ impl RpcClient {
                     .map_ok(|(region, client)| {
                         (scan, region.range(), RawContext::new(region, client, cf))
                     })
-                    .and_then(|(mut scan, region_range, context)| {
-                        let (start_key, end_key) = scan.range();
+                    .and_then(|(scan, region_range, context)| {
+                        let (ref start_key, ref end_key) = scan.range;
```
Comment (Hoverbear) [inline src/rpc/client.rs]: Any way we can take here to avoid the clone?

Comment (nrc) [reply src/rpc/client.rs]: I think we can, but it requires some significant refactoring of `ScanRegionsContext`. That will clash quite a lot with https://github.com/tikv/client-rust/pull/70, so I'd prefer to leave it until after we land both these PRs. I would like to do that refactoring though.

File: src/kv.rs [hunk 3]
```diff
@@ -320,171 +350,179 @@ impl fmt::Debug for KvPair {
...
-/// let from_range_full = range_full.into_bounds();
-/// assert_eq!(
-///     (Bound::Unbounded, Bound::Unbounded),
-///     from_range_full
+///     (Bound::Included(Key::from("Rust")), Bound::Unbounded),
 /// );
 /// ```
 ///
-/// **But, you should not need to worry about all this:** Many functions accept a `impl KeyRange`
+/// **But, you should not need to worry about all this:** Many functions accept a `impl Into<BoundRange>`
```
Comment (brson) [inline src/kv.rs]:
```suggestion
/// **But, you should not need to worry about all this:** Most functions which operate
/// on ranges will accept any types  which implement `Into<BoundRange>`.
```

Comment (brson) [reply src/kv.rs]: This gets rid of "a impl" by repeating the more clearly written line from the first paragraph.

**Pull request merged**